Federation: Removing duplicate finalizer manipulation logic in federation controllers

nikhiljindal opened the pull request:
Ref #40989

Using apimachinery's meta accessor to manipulate objects instead of duplicating that logic in each controller.
Just a cleanup, no behavior change.

cc @kubernetes/sig-federation-pr-reviews

nikhiljindal committed:
Using apimachinerymeta to manipulate finalizers to remove duplicate code

nikhiljindal committed:
Autogenerated bazel changes

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44139)
<!-- Reviewable:end -->

marun commented:
Nice. I only recently learned of meta.Accessor, one less thing for a type adapter to worry about.

nikhiljindal commented:
> Nice. I only recently learned of meta.Accessor, one less thing for a type adapter to worry about.

Yup. This accessor is pretty cool! Helps us remove so much duplicate code :)

nikhiljindal commented:
Rebased and verified that e2e tests pass

marun commented:
I think this PR is a good opportunity to provide unit testing of the code being changed.  Integration and e2e testing isn't good for validating corner cases.

---

Reviewed 16 of 16 files at r1.
Review status: 6 of 15 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 197 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8k-Spqa8xp7qH_jeo:-Kh8k-Spqa8xp7qH_jep:b-u5icws) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L197)):*
> ```Go
> func (dh *DeletionHelper) addFinalizers(obj runtime.Object, finalizers []string) (runtime.Object, error) {
> 	isUpdated, err := finalizersutil.AddFinalizers(obj, finalizers)
> 	if err != nil {
> ```

Consider collapsing these 2 conditions (e.g. ``if err != nil || !isUpdated`` ) since ``return obj err`` is the result for both.

Also, if the functions in ``finalizersutil`` were rewritten to use ``sets.String``, they might be simple enough to inline.

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 215 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8mZO0zZVDUojvS8ef:-Kh8mZO0zZVDUojvS8eg:b-7ropln) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L215)):*
> ```Go
> func (dh *DeletionHelper) removeFinalizers(obj runtime.Object, finalizers []string) (runtime.Object, error) {
> 	isUpdated, err := finalizersutil.RemoveFinalizers(obj, finalizers)
> 	if err != nil {
> ```

ditto

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 1 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8dG9-YrSViIRyFcSN:-Kh8dG9-YrSViIRyFcSO:bliq2sa) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L1)):*
> ```Go
> /*
> ```

Consider adding unit tests for these functions now that there is only one copy of them.  I know they're validated at a higher level, but coverage at this level will simplify maintenance and allow trivial validation of corner cases.

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 26 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8dmQ4Qs4fHI6yYe5q:-Kh8dmQ4Qs4fHI6yYe5r:b-9ilcnj) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L26)):*
> ```Go
> )
> 
> // Returns true if the given object has the given finalizer in its ObjectMeta.
> ```

Consider following the golang comment convention of 'MethodName does this' for exported functions

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 27 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8kc0Sy0Di-GA9SE4V:-Kh8kc0Sy0Di-GA9SE4W:bxp5qb6) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L27)):*
> ```Go
> 
> // Returns true if the given object has the given finalizer in its ObjectMeta.
> func HasFinalizer(obj runtime.Object, finalizer string) (bool, error) {
> ```

Consider rewriting with ``sets.String``:

```golang
finalizers := sets.NewString(accessor.GetFinalizers()...)
return finalizers.Has(finalizer), nil
``` 

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 48 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8eyIqwFqP-UCcLyKB:-Kh8eyIqwFqP-UCcLyKC:bl6obth) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L48)):*
> ```Go
> 		return false, err
> 	}
> 	oldFinalizers := accessor.GetFinalizers()
> ```

Consider rewriting with ``sets.String``:

```golang
// import "k8s.io/apimachinery/pkg/util/sets"

oldFinalizers := sets.NewString(accessor.GetFinalizers()...)
// Better yet, have the caller provide finalizers as an instance of sets.String
newFinalizers := sets.NewString(finalizers...)
if oldFinalizers.IsSuperset(newFinalizers) {
  return false, nil
}
allFinalizers := oldFinalizers.Union(newFinalizers)
access.SetFinalizers(allFinalizers.List())
return true, nil
```

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 63 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8iS8RrDQJmWLmHzcU:-Kh8iS8RrDQJmWLmHzcV:bcsrmvh) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L63)):*
> ```Go
> }
> 
> // Removes the given finalizers from the given objects ObjectMeta.
> ```

Consider rewriting with ``sets.String``

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-:-Kh8mzqW5HenFuXfiMkQ:b70u5wo)*
<!-- Sent from Reviewable.io -->

nikhiljindal committed:
Updating tests

nikhiljindal commented:
Review status: 0 of 22 files reviewed at latest revision, 7 unresolved discussions.

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 197 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8k-Spqa8xp7qH_jeo:-KhAujAsFz3urlfyD_nS:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L197)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider collapsing these 2 conditions (e.g. ``if err != nil || !isUpdated`` ) since ``return obj err`` is the result for both.

Also, if the functions in ``finalizersutil`` were rewritten to use ``sets.String``, they might be simple enough to inline.
</blockquote></details>

Done

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 215 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8mZO0zZVDUojvS8ef:-KhAuokDHyDpQcgKjWn4:b-tl652u) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L215)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

ditto :)

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 1 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8dG9-YrSViIRyFcSN:-KhAur91qWDXoNCrHdkZ:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L1)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider adding unit tests for these functions now that there is only one copy of them.  I know they're validated at a higher level, but coverage at this level will simplify maintenance and allow trivial validation of corner cases.
</blockquote></details>

done

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 26 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8dmQ4Qs4fHI6yYe5q:-KhAutWlh3OcYAwNVaeV:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L26)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider following the golang comment convention of 'MethodName does this' for exported functions
</blockquote></details>

done

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 27 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8kc0Sy0Di-GA9SE4V:-KhAuuMO5xdXF1YTPoXq:b-wuonwa) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L27)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider rewriting with ``sets.String``:

```golang
finalizers := sets.NewString(accessor.GetFinalizers()...)
return finalizers.Has(finalizer), nil
``` 

</blockquote></details>

done. Thanks for the great suggestion!

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 48 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8eyIqwFqP-UCcLyKB:-KhAuzZVLU-AS0geSGaU:bx83mwn) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L48)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider rewriting with ``sets.String``:

```golang
// import "k8s.io/apimachinery/pkg/util/sets"

oldFinalizers := sets.NewString(accessor.GetFinalizers()...)
// Better yet, have the caller provide finalizers as an instance of sets.String
newFinalizers := sets.NewString(finalizers...)
if oldFinalizers.IsSuperset(newFinalizers) {
  return false, nil
}
allFinalizers := oldFinalizers.Union(newFinalizers)
access.SetFinalizers(allFinalizers.List())
return true, nil
```
</blockquote></details>

done. Also changed the callers to send sets.

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 63 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-Kh8iS8RrDQJmWLmHzcU:-KhAv-TlTicRQCrtvg-l:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/a150bc7d85c8a1c72db39d9d202a344dd06feb00/federation/pkg/federation-controller/util/finalizers/finalizers.go#L63)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider rewriting with ``sets.String``
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Updated code as per comments

marun commented:
Thank you for cleanup.  Some more comments inline.  Mostly suggestions, but the use of ``assert`` instead of ``require`` will need to be fixed before merge.  

Also, any reason to keep requiring ``objNameFunc`` vs using meta.Accessor() for that case too?  Or are you intending to tackle that separately?

---

Reviewed 17 of 22 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/federation-controller/ingress/ingress_controller_test.go, line 336 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOgOXndoqKFDWmi5r5:-KhOgOXndoqKFDWmi5r6:bqglrdz) ([raw file](https://github.com/kubernetes/kubernetes/blob/8b2782a31100f407391dcaa3dce3c31bc2049b9e/federation/pkg/federation-controller/ingress/ingress_controller_test.go#L336)):*
> ```Go
> 		}
> 		ingress := obj.(*extensionsv1beta1.Ingress)
> 		if ingressController.hasFinalizerFunc(ingress, metav1.FinalizerOrphanDependents) &&
> ```

This logic appears to duplicate ``DeletionHelper.EnsureFinalizers``.  Consider ensuring that the logic is shared.

---

*[federation/pkg/federation-controller/namespace/namespace_controller_test.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOPk0F0y2976wiKGtC:-KhOPk0F0y2976wiKGtD:b-5dgqu7) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/namespace/namespace_controller_test.go#L103)):*
> ```Go
> 	// Verify that the DeleteFromUnderlyingClusters finalizer is added to the namespace.
> 	updatedNamespace := GetNamespaceFromChan(namespaceUpdateChan)
> 	assert.NotNil(t, updatedNamespace)
> ```

``assert`` methods only log errors, they don't fail fast.  If you want to protect subsequent statements against a nil namespace, you'll need to use methods from the ``require`` package.

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 87 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOTotuf1ottoz8X0n8:-KhOTotuf1ottoz8X0n9:b4z5myq) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L87)):*
> ```Go
> 	runtime.Object, error) {
> 	finalizers := sets.String{}
> 	hasFinalizer, err := finalizersutil.HasFinalizer(obj, FinalizerDeleteFromUnderlyingClusters)
> ```

Consider inlining dh.addFinalizers and finalizersutils.AddFinalizers.  There are no other callers of those functions, and the logic could be simplified to something like the following:

```golang
// Consider creating a function to return this rather than having to define it separately for EnsureFinalizers and HandleObjectInUnderlyingClusters
finalizers := sets.String(FinalizerDeleteFromUnderlyingClusters, metav1.FinalizerOrphanDependents)

accessor, err := meta.Accessor(obj)
oldFinalizers := sets.NewString(accessor.GetFinalizers()...)
if oldFinalizers.IsSuperset(finalizers) {
 return
}

addedFinalizers := finalizers.Difference(oldFinalizers)
glog.V(2).Infof("Adding finalizers %v to %s", addedFinalizers.List(), accessor.GetName())

allFinalizers := oldFinalizers.Union(finalizers)
accessor.SetFinalizers(allFinalizers.List())
updatedObj, err := dh.updateObjFunc(obj)
```

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 89 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOR4fsvZwwSDvaGcWe:-KhOR4fsvZwwSDvaGcWf:baad93n) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L89)):*
> ```Go
> 	hasFinalizer, err := finalizersutil.HasFinalizer(obj, FinalizerDeleteFromUnderlyingClusters)
> 	if err != nil {
> 		return obj, err
> ```

If the intention is to return values from this method, the signature will need to reflect that.  I'm assuming the reason you missed that is because there is no unit test coverage of the deletion helper and the e2e tests are expensive to run.  Consider trying out the integration tests, which provide some coverage for the deletion helper:

 - Ensure etcd is installed and in the path
   - https://github.com/kubernetes/community/blob/master/contributors/devel/testing.md#install-etcd-dependency
 - Run the integration tests
   - make test-integration WHAT=./test/integration/federation

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 192 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOZzUQmxaFa_sNtGl6:-KhOZzUQmxaFa_sNtGl7:b-bsh3et) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L192)):*
> ```Go
> 
> 	// All done. Just remove the finalizer.
> 	return dh.removeFinalizers(obj, sets.NewString(FinalizerDeleteFromUnderlyingClusters))
> ```

Consider changing removeFinalizers to accept just the object and have the set of finalizers to remove be constant.

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 210 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhO_uE-8tL6dXgHo0RE:-KhO_uE-8tL6dXgHo0RF:b-tjmkr7) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L210)):*
> ```Go
> 
> // Removes the given finalizers from the given objects ObjectMeta.
> func (dh *DeletionHelper) removeFinalizers(obj runtime.Object, finalizers sets.String) (runtime.Object, error) {
> ```

Similar suggestion regarding inlining finalizersutil.RemoveFinalizers since this function is the only caller.

---

*[federation/pkg/federation-controller/util/finalizers/finalizers.go, line 60 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOavLpnmcFcWBZ6vCr:-KhOavLpnmcFcWBZ6vCs:bjlxx7l) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers.go#L60)):*
> ```Go
> 	}
> 	oldFinalizers := sets.NewString(accessor.GetFinalizers()...)
> 	if oldFinalizers.Intersection(finalizers).Len() == 0 {
> ```

(No action required) Consider computing the difference and comparing the length of the old and new finalizers instead of performing an intersection and a difference separately.

---

*[federation/pkg/federation-controller/util/finalizers/finalizers_test.go, line 37 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOb8dQMVNcSYlPCe_2:-KhOb8dQMVNcSYlPCe_3:b-xbxpna) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers_test.go#L37)):*
> ```Go
> 
> func TestHasFinalizer(t *testing.T) {
> 	testCases := []struct {
> ```

(No action required) Some suggestions for improvement:

```golang
testCases := map[string]struct {
    objFinalizers []string
    ...
} {
    // The name of a test is an opportunity to document your intent
    "Degenerate case - empty finalizer list won't have null finalizer": {},
    "Return false if the specified finalizer is not in the object's list": {
      objFinalizers: []string{"someFinalizer"},
      // Where it doesn't compromise clarity, consider skipping the definition of type defaults
    },
}

For testName, testCase := range testCases {
    // Use golang's subtest capability to enable parallel execution and filtering on test cases
    //  e.g. make test WHAT=./federation/pkg/federation-controller/util/finalizers KUBE_GOFLAGS="-v" KUBE_TEST_ARGS="-run ^TestHasFinalizer/Degenerate"
    t.Run(testName, func (t *testing.T) {
        // Where possible, do the work in the test.  It's simpler to do it once rather than for every test case
        obj := newObj(testCase.objFinalizers)
        ...
    }
}
    
```


---

*[federation/pkg/federation-controller/util/finalizers/finalizers_test.go, line 38 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhObipV1dpYWOBmzlgy:-KhObipWgebRR23v8J3R:brzi60a) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers_test.go#L38)):*
> ```Go
> func TestHasFinalizer(t *testing.T) {
> 	testCases := []struct {
> 		Obj       runtime.Object
> ```

(No action required) Why are these publicly scoped?

---

*[federation/pkg/federation-controller/util/finalizers/finalizers_test.go, line 75 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOdyziipQBNKCUbzRj:-KhOdyziipQBNKCUbzRk:bdmkhm0) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers_test.go#L75)):*
> ```Go
> 	for index, test := range testCases {
> 		hasFinalizer, _ := HasFinalizer(test.Obj, test.Finalizer)
> 		assert.Equal(t, hasFinalizer, test.Result, fmt.Sprintf("Test case %d failed. Expected: %v, actual: %v", index, test.Result, hasFinalizer))
> ```

(No action required) Since there is no subsequent testing, ``assert`` is safe, but I recommend using ``requires`` by default since it's fail-fast behavior is what most people will expect.

---

*[federation/pkg/federation-controller/util/finalizers/finalizers_test.go, line 119 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOeR2RTxmpuepx6zCD:-KhOeR2RTxmpuepx6zCE:b-r4ql6b) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers_test.go#L119)):*
> ```Go
> 	for index, test := range testCases {
> 		isUpdated, _ := AddFinalizers(test.Obj, test.Finalizers)
> 		assert.Equal(t, isUpdated, test.IsUpdated, fmt.Sprintf("Test case %d failed. Expected isUpdated: %v, actual: %v", index, test.IsUpdated, isUpdated))
> ```

As per previous comments, ``assert`` methods won't fail the test, and ``require`` should be used if that is your intent.

---

*[federation/pkg/federation-controller/util/test/test_helper.go, line 358 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOfrTTTDWy9n5HiiVN:-KhOfrTTTDWy9n5HiiVO:b-d3s2mw) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/test/test_helper.go#L358)):*
> ```Go
> func AssertHasFinalizer(t *testing.T, obj runtime.Object, finalizer string) {
> 	hasFinalizer, err := finalizersutil.HasFinalizer(obj, finalizer)
> 	assert.Nil(t, err)
> ```

``assert`` -> ``requires``

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-:-KhOQYmbCiUUCAY02FRS:bkzl43u)*
<!-- Sent from Reviewable.io -->

marun commented:
@perotinus You may be interested in the suggestions for unit test best practices.

nikhiljindal commented:
Updated assert to require where it made sense.
Can send another PR for objNameFunc

---

Review status: 9 of 23 files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/federation-controller/namespace/namespace_controller_test.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOPk0F0y2976wiKGtC:-KiEJ9FjLuWxzMa4zjhZ:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/namespace/namespace_controller_test.go#L103)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

``assert`` methods only log errors, they don't fail fast.  If you want to protect subsequent statements against a nil namespace, you'll need to use methods from the ``require`` package.
</blockquote></details>

done

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 87 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOTotuf1ottoz8X0n8:-KiEJCPDnbLTKMzQHhRK:b-r6j0iu) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L87)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider inlining dh.addFinalizers and finalizersutils.AddFinalizers.  There are no other callers of those functions, and the logic could be simplified to something like the following:

```golang
// Consider creating a function to return this rather than having to define it separately for EnsureFinalizers and HandleObjectInUnderlyingClusters
finalizers := sets.String(FinalizerDeleteFromUnderlyingClusters, metav1.FinalizerOrphanDependents)

accessor, err := meta.Accessor(obj)
oldFinalizers := sets.NewString(accessor.GetFinalizers()...)
if oldFinalizers.IsSuperset(finalizers) {
 return
}

addedFinalizers := finalizers.Difference(oldFinalizers)
glog.V(2).Infof("Adding finalizers %v to %s", addedFinalizers.List(), accessor.GetName())

allFinalizers := oldFinalizers.Union(finalizers)
accessor.SetFinalizers(allFinalizers.List())
updatedObj, err := dh.updateObjFunc(obj)
```
</blockquote></details>

keeping set manipulation and calling updateObjFunc separate allows for simpler unit testing.

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 89 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOR4fsvZwwSDvaGcWe:-KiEJQOywjPQKFzOoygV:byk16fh) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L89)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

If the intention is to return values from this method, the signature will need to reflect that.  I'm assuming the reason you missed that is because there is no unit test coverage of the deletion helper and the e2e tests are expensive to run.  Consider trying out the integration tests, which provide some coverage for the deletion helper:

 - Ensure etcd is installed and in the path
   - https://github.com/kubernetes/community/blob/master/contributors/devel/testing.md#install-etcd-dependency
 - Run the integration tests
   - make test-integration WHAT=./test/integration/federation
</blockquote></details>

um an not sure why you think the signature does not reflect that?

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 210 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhO_uE-8tL6dXgHo0RE:-KiEJfjPhJJfbmvpqbyO:bnj7hl0) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L210)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Similar suggestion regarding inlining finalizersutil.RemoveFinalizers since this function is the only caller.
</blockquote></details>

same as above

---

*[federation/pkg/federation-controller/util/finalizers/finalizers_test.go, line 38 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhObipV1dpYWOBmzlgy:-KiEJlvQsHQqWaxaAkrF:b-7ql1td) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers_test.go#L38)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Why are these publicly scoped?
</blockquote></details>

fixed

---

*[federation/pkg/federation-controller/util/finalizers/finalizers_test.go, line 75 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOdyziipQBNKCUbzRj:-KiEJoPWquuTyftHxyKp:brz4ng) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers_test.go#L75)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Since there is no subsequent testing, ``assert`` is safe, but I recommend using ``requires`` by default since it's fail-fast behavior is what most people will expect.
</blockquote></details>

keeping to assert so that single test run produces information about all the failing cases.

---

*[federation/pkg/federation-controller/util/finalizers/finalizers_test.go, line 119 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOeR2RTxmpuepx6zCD:-KiEJyZhn081wI4IGyvT:bnj7hl0) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers_test.go#L119)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

As per previous comments, ``assert`` methods won't fail the test, and ``require`` should be used if that is your intent.
</blockquote></details>

same as above

---

*[federation/pkg/federation-controller/util/test/test_helper.go, line 358 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOfrTTTDWy9n5HiiVN:-KiEK0GpHKVE8GRNF3Uj:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/test/test_helper.go#L358)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

``assert`` -> ``requires``
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-:-KiEIrEa8i5j6WjMYbj7:bkzvx9n)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: 9 of 23 files reviewed at latest revision, 12 unresolved discussions.

---

*[federation/pkg/federation-controller/namespace/namespace_controller_test.go, line 103 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOPk0F0y2976wiKGtC:-KiEKGyd83nDQy4OdzSX:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/namespace/namespace_controller_test.go#L103)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

done
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Thanks for the detailed review @marun 
Updated code as per comments.
PTAL

marun commented:
Reviewed 4 of 15 files at r4, 6 of 6 files at r5.
Review status: 19 of 23 files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federation-controller/sync/controller.go, line 210 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KiHA5YBTcSkCmlrLymq:-KiHA5YBTcSkCmlrLymr:b-tnr69a) ([raw file](https://github.com/kubernetes/kubernetes/blob/1ebd4bfdf5eacaa75c5f63c12037a6265da7db80/federation/pkg/federation-controller/sync/controller.go#L210)):*
> ```Go
> // Sends the given updated object to apiserver.
> // Assumes that the given object is a secret.
> func (s *FederationSyncController) updateSecret(obj pkgruntime.Object) (pkgruntime.Object, error) {
> ```

s/Secret/Object/

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 87 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOTotuf1ottoz8X0n8:-KiH3Zqgvue24g-Ao2b3:bc2mdy8) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L87)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

keeping set manipulation and calling updateObjFunc separate allows for simpler unit testing.
</blockquote></details>

Computing the new set of finalizers is separable from updating them, and the proposed approach spreads that computation across 2 functions.  I agree that separating out the call to the update function can be a win (less dependency), so how about the following?

```golang
updatedObj, err := setFinalizers(obj)
if err {
  // Not sure why the existing function insists on always returning the object.  Most golang code I've seen doesn't expect a value to be returned alongside an error.
  return nil, err
}
// setFinalizers would return nil if the desired finalizers were already present
if updatedObj != nil {
  return dh.updateObjFunc(updatedObj)
}
```

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 89 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOR4fsvZwwSDvaGcWe:-KiH4b9l6hUIDS-DZS1n:b-5nudsv) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L89)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

um an not sure why you think the signature does not reflect that?
</blockquote></details>

You're right.  My brain was misreading ``(`` as ``{``.

---

*[federation/pkg/federation-controller/util/finalizers/finalizers_test.go, line 75 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOdyziipQBNKCUbzRj:-KiH9c257NO1UuJOgS1q:bbu26d7) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/finalizers/finalizers_test.go#L75)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

keeping to assert so that single test run produces information about all the failing cases.
</blockquote></details>

(No action required) I think it would be wise to stop requiring soft asserts to ensure execution of subsequent test cases.  golang subtests provide the same guarantee without the potential for confusing maintainers as to whether a hard or soft failure is expected. 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: 19 of 23 files reviewed at latest revision, 8 unresolved discussions.

---

*[federation/pkg/federation-controller/sync/controller.go, line 210 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KiHA5YBTcSkCmlrLymq:-KiNOktGh1wazVqlBJ0h:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/1ebd4bfdf5eacaa75c5f63c12037a6265da7db80/federation/pkg/federation-controller/sync/controller.go#L210)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

s/Secret/Object/
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 87 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOTotuf1ottoz8X0n8:-KiNO93TnWAyDrg52dXM:bysgo39) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L87)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Computing the new set of finalizers is separable from updating them, and the proposed approach spreads that computation across 2 functions.  I agree that separating out the call to the update function can be a win (less dependency), so how about the following?

```golang
updatedObj, err := setFinalizers(obj)
if err {
  // Not sure why the existing function insists on always returning the object.  Most golang code I've seen doesn't expect a value to be returned alongside an error.
  return nil, err
}
// setFinalizers would return nil if the desired finalizers were already present
if updatedObj != nil {
  return dh.updateObjFunc(updatedObj)
}
```
</blockquote></details>

Its still not clear to me that how can we share that setFinalizers in EnsureFinalizers and HandleObjectInUnderlyingClusters. They are doing different things.

set manipulation and call to update function are already separate.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Thanks for the review @marun
Updated code as per comments.
Ready for another round :)

marun commented:
/lgtm

---

Reviewed 4 of 15 files at r4, 1 of 7 files at r6, 6 of 6 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go, line 87 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-KhOTotuf1ottoz8X0n8:-KiWnbwSQLEbpZRLJW0s:bg6zcdd) ([raw file](https://github.com/kubernetes/kubernetes/blob/509fe2808cf28bf532ce932f9f975cf238b9e9b5/federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go#L87)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Its still not clear to me that how can we share that setFinalizers in EnsureFinalizers and HandleObjectInUnderlyingClusters. They are doing different things.

set manipulation and call to update function are already separate.
</blockquote></details>

I wasn't suggesting sharing setFinalizers between removeFinalizers and EnsureFinalizers, this was just for EnsureFinalizers.  

In any case, what you've proposed works regardless of my preference a slightly simpler solution.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44139#-:-KiWvy28DyyYZR2khEPj:b-8ks5qa)*
<!-- Sent from Reviewable.io -->

marun commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

nikhiljindal commented:
Prow is showing me SUCCESS but looking at the logs, one test case `Federated Services [Feature:Federation] with clusters DNS should be able to discover a federated service` failed (timed out). Filed https://github.com/kubernetes/test-infra/issues/2575 for that.

Retrying @k8s-bot pull-kubernetes-federation-e2e-gce test this

madhusudancs commented:
If the gubernator dashboard for the run says `SUCCESS`, please feel free to merge the PR. I guess you are referring to this run - https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/44139/pull-kubernetes-federation-e2e-gce/2090/

We set failure/flake tolerance to 2 in our presubmits. This runs the failing tests the second time to ensure that the submit queue is not slowed down by random flakes. If you see the same test failing twice, then that's a problem worth debugging. If you see a failure only once, then it is probably Ok in most cases, but use caution.

nikhiljindal commented:
got it. Thats the same reply I got on https://github.com/kubernetes/test-infra/issues/2575
